roachtest: add limit_capacity roachtests

kvoli opened the pull request:
There were no roachtests which limited the capacity of a node in a
cluster where better allocation decisions would have an impact.
Introduce two roachtests

- `limit_capacity/write_b=25mbs`
- `limit_capacity/compact=0`

The test uses a 4 node cluster running a fixed rate kv50 workload
consuming 20-30% CPU and approx 100mb/s write bandwidth after 10
minutes. After 10 minutes, a node's capacity is limited.

In `write_b=25mbs`, the write bandwidth is limited to 25mb/s, or 1/4 of
the current write bandwidth usage. In `compact=0`, non-flush compactions
are disabled.

These tests are manual only and should be enabled either once the work
needed to maintain the steady state QPS is merged, or without assertions
using roachperf.

Resolves: https://github.com/cockroachdb/cockroach/issues/116640
Resolves: https://github.com/cockroachdb/cockroach/issues/116823

Release note: None

kvoli committed:
builtins: allow setting compaction concurrency to zero

Setting `SELECT
crdb_internal.set_compaction_concurrency(nodeID,storeID,0)` is now
allowed. Setting the compaction concurrency of a store to zero will
disable compactions on that store. This change could be gated behind a
cluster setting, however the likelihood of setting a value of zero
accidentally is minimal. Values less than zero will still return an
error.

Epic: none
Release note: None

kvoli committed:
roachtest: generalize lease preferences in configure zones

Previously, `configureZones` enabled setting a preferred leaseholder
node but not preferences generally. The function is useful generally as
it composes with zone config application to multiple targets. Update the
lease preferences to more generally take a string.

Part of: #116823
Release note: None

kvoli committed:
roachtest: add limit_capacity roachtests

There were no roachtests which limited the capacity of a node in a
cluster where better allocation decisions would have an impact.
Introduce two roachtests

- `limit_capacity/write_b=25mbs`
- `limit_capacity/compact=0`

The test uses a 4 node cluster running a fixed rate kv50 workload
consuming 20-30% CPU and approx 100mb/s write bandwidth after 10
minutes. After 10 minutes, a node's capacity is limited.

In 'write_b=25mbs', the write bandwidth is limited to 25mb/s, or 1/4 of
the current write bandwidth usage. In `compact=0`, non-flush compactions
are disabled.

These tests are manual only and should be enabled either once the work
needed to maintain the steady state QPS is merged, or without assertions
using roachperf.

Resolves: #116640
Resolves: #116823

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/117702)
<!-- Reviewable:end -->

kvoli commented:
NB: I've added some lease preference roachtest cleanup here in order to later use the refactor to remove leases from the node which has limited capacity.

I removed these from the initial PR here as they didn't serve a purpose yet, without leases the QPS will remain at 100% regardless of limiting the capacity on the follower only node.

aadityasondhi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117702#-:-NomJ1xJ2P-yQZobXaz6:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist, @herkolategan, @kvoli, @srosenberg, and @sumeerbhola)

___
*[`pkg/cmd/roachtest/tests/limit_capacity.go` line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/117702#-NomIojm378q16ZNmH4o:-NomIojm378q16ZNmH4p:b1nbglb) ([raw file](https://github.com/cockroachdb/cockroach/blob/99c7ec0aba0251ef09b4c8f9a808783f481c2e80/pkg/cmd/roachtest/tests/limit_capacity.go#L110)):*
> ```Go
> 
> 	t.Status(fmt.Sprintf("waiting %s for baseline workload throughput", initialDuration))
> 	wait(c.NewMonitor(ctx, nodes), limitDuration)
> ```

I believe you want `initialDuration` here.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
TYFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117702#-:-NomrOqZ3mkpiuzNV3Ly:bxgjrxj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi, @andrewbaptist, @herkolategan, @srosenberg, and @sumeerbhola)

___
*[`pkg/cmd/roachtest/tests/limit_capacity.go` line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/117702#-NomIojm378q16ZNmH4o:-NomrJRjDnZU8nn3OBCt:b-f85ewl) ([raw file](https://github.com/cockroachdb/cockroach/blob/99c7ec0aba0251ef09b4c8f9a808783f481c2e80/pkg/cmd/roachtest/tests/limit_capacity.go#L110)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

I believe you want `initialDuration` here.
</blockquote></details>

Good catch, updated to `initialDuration`. 


<!-- Sent from Reviewable.io -->

kvoli commented:
bors r=aadityasondhi

yuzefovich commented:
I think bors complains about this PR:
```
[23:43:04][Run unit tests] ERROR: /go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/BUILD.bazel:4:11: GoCompilePkg pkg/cmd/roachtest/tests/tests.a failed: (Exit 1): builder failed: error executing command (from target //pkg/cmd/roachtest/tests:tests) bazel-out/k8-opt-exec-2B5CBBC6/bin/external/go_sdk/builder_reset/builder compilepkg -sdk external/go_sdk -installsuffix linux_amd64 -tags bazel,gss,bazel,gss -src pkg/cmd/roachtest/tests/acceptance.go ... (remaining 590 arguments skipped)
[23:43:04][Run unit tests] 
[23:43:04][Run unit tests] Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
[23:43:04][Run unit tests] pkg/cmd/roachtest/tests/limit_capacity.go:93:13: cannot use appNode (variable of type "github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option".NodeListOption) as install.RunOptions value in argument to c.Run
[23:43:04][Run unit tests] pkg/cmd/roachtest/tests/limit_capacity.go:102:22: cannot use appNode (variable of type "github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option".NodeListOption) as install.RunOptions value in argument to c.RunE
[23:43:04][Run unit tests] pkg/cmd/roachtest/tests/limit_capacity.go:115:14: cannot use limitedNode (variable of type "github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option".NodeListOption) as install.RunOptions value in argument to c.Run
[23:43:04][Run unit tests] compilepkg: error running subcommand external/go_sdk/pkg/tool/linux_amd64/compile: exit status 2
```

bors r-

kvoli commented:
> I think bors complains about this PR:
> 
> ```
> [23:43:04][Run unit tests] ERROR: /go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/BUILD.bazel:4:11: GoCompilePkg pkg/cmd/roachtest/tests/tests.a failed: (Exit 1): builder failed: error executing command (from target //pkg/cmd/roachtest/tests:tests) bazel-out/k8-opt-exec-2B5CBBC6/bin/external/go_sdk/builder_reset/builder compilepkg -sdk external/go_sdk -installsuffix linux_amd64 -tags bazel,gss,bazel,gss -src pkg/cmd/roachtest/tests/acceptance.go ... (remaining 590 arguments skipped)
> [23:43:04][Run unit tests] 
> [23:43:04][Run unit tests] Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
> [23:43:04][Run unit tests] pkg/cmd/roachtest/tests/limit_capacity.go:93:13: cannot use appNode (variable of type "github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option".NodeListOption) as install.RunOptions value in argument to c.Run
> [23:43:04][Run unit tests] pkg/cmd/roachtest/tests/limit_capacity.go:102:22: cannot use appNode (variable of type "github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option".NodeListOption) as install.RunOptions value in argument to c.RunE
> [23:43:04][Run unit tests] pkg/cmd/roachtest/tests/limit_capacity.go:115:14: cannot use limitedNode (variable of type "github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option".NodeListOption) as install.RunOptions value in argument to c.Run
> [23:43:04][Run unit tests] compilepkg: error running subcommand external/go_sdk/pkg/tool/linux_amd64/compile: exit status 2
> ```
> 
> bors r-

Thanks for cancelling, I'll sort out the skew.

kvoli commented:
bors r=aadityasondhi
